### PR TITLE
Add modal state provider, replace usage except methods

### PR DESCRIPTION
**Pull request by estrattonbailey**

### CODE REVIEW

File: src/state/modals/index.tsx [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+      setActiveModals(activeModals => [...activeModals, modal])
+      setIsModalActive(true)
+    },
+    [setIsModalActive, setActiveModals],
+  )
+
+  const closeModal = React.useCallback(() => {
+    let totalActiveModals = 0
+    setActiveModals(activeModals => {
+      activeModals.pop()
```
Comment (gaearon) [inline src/state/modals/index.tsx:237]: This is mutative. We shouldn't mutate inside updaters, which need to be pure functions.

We should enable Strict Mode to catch these better.

File: src/state/modals/index.tsx [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+const ModalControlContext = React.createContext<{
+  openModal: (modal: Modal) => void
+  closeModal: () => void
+}>({
+  openModal: () => {},
+  closeModal: () => {},
+})
+
+export function Provider({children}: React.PropsWithChildren<{}>) {
+  const [isModalActive, setIsModalActive] = React.useState(false)
```
Comment (gaearon) [inline src/state/modals/index.tsx:222]: I think maybe we can replace this with `const isModalActive = activeModals.length > 0`?

File: src/state/modals/index.tsx [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+      activeModals.pop()
+      totalActiveModals = activeModals.length
+      return activeModals
+    })
+    setIsModalActive(totalActiveModals > 0)
+  }, [setIsModalActive, setActiveModals])
+
+  return (
+    <ModalContext.Provider
+      value={{
```
Comment (gaearon) [inline src/state/modals/index.tsx:246]: Objects we pass to providers should be memoized, let's useMemo them.

File: src/state/modals/index.tsx [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        isModalActive,
+        activeModals,
+      }}>
+      <ModalControlContext.Provider
+        value={{
```
Comment (gaearon) [inline src/state/modals/index.tsx:251]: Same.

File: src/state/models/ui/shell.ts [hunk 5]
```diff
@@ -319,10 +118,11 @@ export class ShellUiModel {
       this.closeLightbox()
       return true
     }
-    if (this.isModalActive) {
-      this.closeModal()
-      return true
-    }
+    // if (this.isModalActive) {
+    //   // TODO
```
Comment (gaearon) [inline src/state/models/ui/shell.ts:122]: In this PR or later?

**Review by gaearon — APPROVED** [review]
nice. accept with small changes

File: src/state/modals/index.tsx (hunk 3 above)
Comment (estrattonbailey) [reply src/state/modals/index.tsx:250]: Good call!

File: src/state/models/ui/shell.ts (hunk 5 above)
Comment (estrattonbailey) [reply src/state/models/ui/shell.ts]: In this PR, just did it last

File: src/state/modals/index.tsx (hunk 1 above)
Comment (estrattonbailey) [reply src/state/modals/index.tsx:241]: Yeah good shout, this should be improved at some point. We should also do the portaling thing we talked about.

This is just a direct port of the existing functionality, so I'd like to just leave as is until we can come back and do it right.

File: src/state/modals/index.tsx (hunk 2 above)
Comment (estrattonbailey) [reply src/state/modals/index.tsx:224]: Sounds like a good idea. Oddly, without it, setting `activeModals` to an empty array doesn't trigger a re-render? Might be a shallow compare issue between arrays, idk. I'm sure we could figure this out but it works right now and the code is clear

File: src/lib/media/picker.web.tsx [hunk 6]
```diff
@@ -14,12 +15,12 @@ export async function openCamera(
...
 
 export async function openCropper(
-  store: RootStoreModel,
+  _store: RootStoreModel,
   opts: CropperOptions,
 ): Promise<RNImage> {
   // TODO handle more opts
   return new Promise((resolve, reject) => {
-    store.shell.openModal({
+    unstable__openModal({
```
Comment (estrattonbailey) [inline src/lib/media/picker.web.tsx:23]: Without this hacky global I'd need to pass through 2 mobx models and into this file 🤔

File: src/state/modals/index.tsx (hunk 1 above)
Comment (pfrazee) [reply src/state/modals/index.tsx:246]: We can make this a non-mutation now though right? Not _terribly_ hard

**Review by pfrazee — APPROVED** [review]
LGTM, I think any further changes can be in followups

**Pull request merged**